### PR TITLE
Remove `AbstractUnnamedTypeDeclaration`

### PR DESCRIPTION
**Pull request by datho7561**
## What it does
See #2481

- Rework hierarchy so that `ImplicitTypeDeclaration` depends on `AbstractTypeDeclaration`
  - Use an empty SimpleName node for `ImplicitTypeDeclaration` as suggested by Stephen
  - Implement `internalResolveBinding()` for `ImplicitTypeDeclaration`
- Revert `AbstractTypeDeclaration` to how it was before `ImplicitTypeDeclaration` was added
- Reincorporate Jörg's fix that makes `typeName` volatile
- Delete `AbstractUnnamedTypeDeclaration`

## How to test
See #2481, it should fix pasting snippets into the package explorer (~~it doesn't yet, it needs work, currently it throws an "Invalid identifier" exception~~).

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by iloveeclipse** [comment]
@jarthana , @stephan-herrmann : tomorrow is RC1 build. It would be very helpful to review (and merge?) this PR ASAP to avoid last minute fixes in RC2.

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/AbstractTypeDeclaration.java [hunk 1]
```diff
@@ -26,14 +28,40 @@
...
-public abstract class AbstractTypeDeclaration extends AbstractUnnamedTypeDeclaration {
+public abstract class AbstractTypeDeclaration extends BodyDeclaration {
 
 	/**
 	 * The type name; lazily initialized; defaults to a unspecified,
 	 * legal Java class identifier.
 	 * @since 2.0 (originally declared on {@link TypeDeclaration})
 	 */
-	volatile SimpleName typeName;
+	volatile SimpleName typeName = null;
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/AbstractTypeDeclaration.java:38]: nitpick: null assignment is not needed

**Comment by srikanth-sankaran** [comment]
> @jarthana , @stephan-herrmann : tomorrow is RC1 build. It would be very helpful to review (and merge?) this PR ASAP to avoid last minute fixes in RC2.

I am told @jarthana is on leave till Monday. I will study this and share my comments. @mpalat has also offered to look into this - Thanks!

**Review by stephan-herrmann — COMMENTED** [review]
Basically looks good, the issue of calling `preLazyInit()`, `postLazyInit()` has priority among my remarks.

Let's wait for at least one more review.

**Comment by stephan-herrmann** [comment]
> See #2481, it should fix pasting snippets into the package explorer (~it doesn't yet, it needs work, currently it throws an "Invalid identifier" exception~).

That requires some reverts in JDT/UI, right?  Should those be prepared already?

**Comment by iloveeclipse** [comment]
> That requires some reverts in JDT/UI, right? Should those be prepared already?

Not reverts but careful changes, as some logic changed too => https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/1420

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ImplicitTypeDeclaration.java [hunk 2]
```diff
@@ -161,4 +161,34 @@ final int internalGetSetIntProperty(SimplePropertyDescriptor property, boolean g
 		return super.internalGetSetIntProperty(property, get, value);
 	}
 
+	// from AbstractTypeDeclaration
+
+	@Override
+	public SimpleName getName() {
+		if (this.typeName == null) {
+			this.typeName = new SimpleName(this.ast) {
```
[the remark this answers is not in the document]
Comment (datho7561) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ImplicitTypeDeclaration.java:169]: I might as well create the static nested class, it doesn't take much more effort.

I don't know what preLazyInit and postLazyInit do, but I'll add them.

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/DefaultBindingResolver.java [hunk 3]
```diff
@@ -1879,6 +1879,25 @@ synchronized ITypeBinding resolveType(TypeDeclaration type) {
 		return null;
 	}
 
+	@Override
+	synchronized ITypeBinding resolveType(ImplicitTypeDeclaration type) {
+		final Object node = this.newAstToOldAst.get(type);
+		if (node instanceof org.eclipse.jdt.internal.compiler.ast.ImplicitTypeDeclaration implicitTypeDeclaration) {
+			IBinding declaringMember = getDeclaringMember(implicitTypeDeclaration, implicitTypeDeclaration.scope);
```
[the remark this answers is not in the document]
Comment (datho7561) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/DefaultBindingResolver.java:1886]: That makes sense to me; I'll replace it with null

**Comment by datho7561** [comment]
Sorry I should have prepared a PR with the required changes to jdt.ui, my bad I'll have that up soon too

**Comment by iloveeclipse** [comment]
> Sorry I should have prepared a PR with the required changes to jdt.ui, my bad I'll have that up soon too

No problem, but you can review https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/1420

**Comment by datho7561** [comment]
Looking into test failures

**Comment by datho7561** [comment]
They are legitimate, working on a fix

**Comment by datho7561** [comment]
I can't reproduce the failure (`JavaSearchBugsTests2.testBug483650_0002`) locally from within Eclipse. I'm going to squash to one commit and clean up the commit message.

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ImplicitTypeDeclaration.java [hunk 4]
```diff
@@ -161,4 +173,43 @@ final int internalGetSetIntProperty(SimplePropertyDescriptor property, boolean g
 		return super.internalGetSetIntProperty(property, get, value);
 	}
 
+	// from AbstractTypeDeclaration
+
+	@Override
+	public SimpleName getName() {
+		if (this.typeName == null) {
+			preLazyInit();
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ImplicitTypeDeclaration.java:181]: I should have looked at the javadoc of preLazyInit() (did you? :smile: ): 
the pattern should always be enclosing in `synchronized(this)`.

**Review by stephan-herrmann — APPROVED** [review]
If build succeeds, this is good to go from my p.o.v.
thanks

**Comment by iloveeclipse** [comment]
I'm going to sleep now.

If you are going to merge this, please also merge https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/1420 at same time, even if it doesn't compile yet on jenkins, it compiles with this PR. Merging only this PR will break the SDK build.

**Comment by mpalat** [comment]
> If build succeeds, this is good to go from my p.o.v. thanks

@stephan-herrmann Since this is still a  preview feature, an "@noreference" to the ImplicitTypeDeclaration is the norm we follow - until it becomes a standard feature. So I would request to add that in the JavaDoc before submitting

**Comment by iloveeclipse** [comment]
@mpalat : I think David can add annotation for the next build. I would like to have all ready for RC1 now, so we can run all tests etc.

**Comment by mpalat** [comment]
> @mpalat : I think David can add annotation for the next build. I would like to have all ready for RC1 now, so we can run all tests etc.

@iloveeclipse - yes, that's sounds fine.

**Pull request merged**